### PR TITLE
nixos: Installer tests for systemd stage 1

### PR DESCRIPTION
**Pull request by ElvishJerricco**
###### Description of changes

Enable *some* of the installer tests for systemd stage 1. We can uncomment the others as we fix their tests.

Currently the *normal* specialisation tests are failing on master the same way that they are for this PR, so I've included them although I haven't verified them.

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by ElvishJerricco** [comment]
I've marked this as a draft because of the `pivot_root` issue in `channel.nix`. If anyone has any idea what's up with that, I'd appreciate some input.

**Comment by ElvishJerricco** [comment]
Oh, and I'm not sure if `echo c > /proc/sysrq-trigger` is the best thing to do in the `panic-on-fail` service.

And finally, the LVM test doesn't work because `boot.initrd.services.lvm.enable` is not enabled by default. Same with `swraid`, though that one appears to fail later anyway. My opinion is that those should be enabled by default so that a person can switch to systemd in stage 1 with zero churn if they currently use mdraid or LVM.

File: nixos/modules/installer/cd-dvd/channel.nix [hunk 1]
```diff
@@ -39,7 +39,8 @@ in
         echo "unpacking the NixOS/Nixpkgs sources..."
         mkdir -p /nix/var/nix/profiles/per-user/root
         ${config.nix.package.out}/bin/nix-env -p /nix/var/nix/profiles/per-user/root/channels \
-          -i ${channelSources} --quiet --option build-use-substitutes false
+          -i ${channelSources} --quiet --option build-use-substitutes false \
+          ${optionalString config.boot.initrd.systemd.enable "--option sandbox false"} # There's an issue with pivot_root
```
Comment (ckiee) [inline nixos/modules/installer/cd-dvd/channel.nix:43]: Also commit-worthy.

File: nixos/tests/installer-systemd-stage-1.nix [hunk 2]
```diff
@@ -0,0 +1,31 @@
+{ system ? builtins.currentSystem
+, config ? {}
+, pkgs ? import ../.. { inherit system config; }
+}:
+
+{
+  inherit (import ./installer.nix { inherit system config pkgs; systemdStage1 = true; })
+    # bcache
```
Comment (ckiee) [inline nixos/tests/installer-systemd-stage-1.nix:8]: Comment why these are commented?

**Review by ckiee — CHANGES_REQUESTED** [review]
Just general QOL things since I don't know much about systemd-in-initrd.

File: nixos/modules/installer/cd-dvd/channel.nix (hunk 1 above)
Comment (ElvishJerricco) [reply nixos/modules/installer/cd-dvd/channel.nix:43]: Ideally this part will be gone and replaced by a separate commit that actually solves the issue, but sure. For now I'll make another commit.

**Comment by lovesegfault** [comment]
Maybe mark as ready to review so people get notified about this?

**Comment by ElvishJerricco** [comment]
I just don't want it to get merged before we've addressed the questions about `pivot_root` and `panic-on-fail`

File: nixos/tests/installer-systemd-stage-1.nix [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    simple
+    simpleLabels
+    simpleProvided
+    simpleSpecialised
+    simpleUefiGrub
+    simpleUefiGrubSpecialisation
+    simpleUefiSystemdBoot
+    # swraid
+    # zfsroot
+    ;
```
Comment (lovesegfault) [inline nixos/tests/installer-systemd-stage-1.nix:31]: https://github.com/NixOS/nixpkgs/projects/51#card-81106305

**Comment by ElvishJerricco** [comment]
Unless anyone has any comments on the `pivot_root` thing, I suggest we merge this.

**Review by lovesegfault — APPROVED** [review]
LGTM

No clue on the `channelSources` thing, adding a note in our TODO list.

File: nixos/modules/installer/cd-dvd/channel.nix (hunk 1 above)
Comment (lovesegfault) [inline nixos/modules/installer/cd-dvd/channel.nix:43]: https://github.com/NixOS/nixpkgs/projects/51#card-81336502

**Pull request merged**